security/certManager/general/certManager.go

```diff
@@ -0,0 +1,224 @@
+	KeyFile                   string `yaml:"keyFile" json:"keyFile" description:"file name of private key in PEM format"`
+	CertFile                  string `yaml:"certFile" json:"certFile" description:"file name of certificate in PEM format"`
+	ClientCertificateRequired bool   `yaml:"clientCertificateRequired" json:"clientCertificateRequired" description:"require client certificate"`
+	UseSystemCAPool           bool   `yaml:"useSystemCAPool" json:"useSystemCAPool" description:"use system certification pool"`
+}
+
+// CertManager holds certificates from filesystem watched for changes
+type CertManager struct {
+	config Config
+
+	watcher                 *fsnotify.Watcher
+	doneWg                  sync.WaitGroup
+	done                    chan struct{}
+	verifyClientCertificate tls.ClientAuthType
+	logger                  *zap.Logger
+
+	mutex      sync.Mutex
+	tlsKeyPair tls.Certificate
+	tlsCAPool  *x509.CertPool
+}
+
+// New creates a new certificate manager which watches for certs in a filesystem
+func New(config Config, logger *zap.Logger) (*CertManager, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	verifyClientCertificate := tls.RequireAndVerifyClientCert
+	if !config.ClientCertificateRequired {
+		verifyClientCertificate = tls.NoClientCert
+	}
+
+	c := &CertManager{
+		watcher:                 watcher,
+		config:                  config,
+		verifyClientCertificate: verifyClientCertificate,
+		logger:                  logger,
+	}
+	err = c.loadCAs()
+	if err != nil {
+		return nil, err
+	}
+	err = c.loadCerts()
+	if err != nil {
+		return nil, err
+	}
+	if err := c.watcher.Add(filepath.Dir(config.CAFile)); err != nil {
+		return nil, err
+	}
+	if err := c.watcher.Add(filepath.Dir(config.CertFile)); err != nil {
+		return nil, err
+	}
+	if err := c.watcher.Add(filepath.Dir(config.KeyFile)); err != nil {
+		return nil, err
+	}
+
+	c.done = make(chan struct{})
+	c.doneWg.Add(1)
+
+	go c.watchFiles()
+
+	return c, nil
+}
+
+// GetCertificateAuthorities returns certificates authorities
+func (a *CertManager) GetCertificateAuthorities() *x509.CertPool {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	return a.tlsCAPool
+}
+
+// GetServerTLSConfig returns tls configuration for servers
+func (a *CertManager) GetServerTLSConfig() *tls.Config {
+	return &tls.Config{
+		ClientCAs:      a.GetCertificateAuthorities(),
+		GetCertificate: a.getServerCertificate,
+		MinVersion:     tls.VersionTLS12,
+		ClientAuth:     a.verifyClientCertificate,
+	}
+}
+
+// GetClientTLSConfig returns tls configuration for clients
+func (a *CertManager) GetClientTLSConfig() *tls.Config {
+	return &tls.Config{
+		RootCAs:                  a.GetCertificateAuthorities(),
+		GetClientCertificate:     a.getClientCertificate,
+		PreferServerCipherSuites: true,
+		MinVersion:               tls.VersionTLS12,
+		InsecureSkipVerify:       true,   // To skip verifying private certificate of server, it should be true
```

Comment:
We want to always verify server certificate. Pls remove the option InsecureSkipVerify.